Infinite scroll in Media

merapi opened the pull request:
This adds infinite scroll to the Media gallery (https://github.com/google/web-stories-wp/issues/23).

Next steps (can be included in this PR):

1. Lazy load images - right now we have `loading="lazy"` which works in [Chrome/Edge, soon FF](https://caniuse.com/#search=lazy%20images), we can utilize IntersectionObserver to build our own solution. 
This solution loads as many pages as needed to fill the whole viewport. For example with per_page=2 and rootMargin: 400px it will load ~10 pages to fill the viewport and then load one row (2 images) as you start scrolling down - not sure if it's useful but it's there "for free" (for per_page=100 it will load 100 one time - as expected in AC).

2. Keyboard support - be able to see all images with arrows.

3. UX, loading states.

merapi committed:
Infinite scroll (paginated) implemented

spacedmonkey commented on assets/src/edit-story/app/api/apiProvider.js:
```diff
@@ -105,10 +105,10 @@ function APIProvider({ children }) {
       let apiPath = media;
-      const perPage = 100;
-      apiPath = addQueryArgs(apiPath, { per_page: perPage });
+      const perPage = 20;
```
Why change to 20? `The library will be paginated (100 per page)` The AC says https://github.com/google/web-stories-wp/issues/23

merapi replied:
Maybe my note was not clear: it's WIP/for easier testing (20 is one viewport on FHD).

spacedmonkey replied:
If it is a WIP in process, put [WIP] in title of PR and set the PR as a draft. Otherwise it will get reviewed and merged.

spacedmonkey commented on assets/src/edit-story/components/library/panes/media/mediaPane.js:
```diff
@@ -182,69 +222,93 @@ function MediaPane(props) {
+                    />
+                  ))}
+              </Column>
+              {hasMore && <div ref={refContainerFooter}>{'Loading...'}</div>}
```
```suggestion
              {hasMore && <div ref={refContainerFooter}>{__('Loading', 'web-stories')}</div>}
```

spacedmonkey commented on assets/src/edit-story/utils/useIntersectionEffect.js:
```diff
@@ -0,0 +1,59 @@
+ * External dependencies
+ */
+import { useEffect } from 'react';
+import IntersectionObserver from 'intersection-observer-polyfill';
```
I am not sure we support ie11 in this project. Other than that broswer, do we need this. 

CC @pbakaus

merapi replied:
For now, I followed [Dima](https://github.com/google/web-stories-wp/blame/c29f388b8135b9b20e061d036c01222f8fa10cbb/assets/src/edit-story/utils/useResizeEffect.js#L20), we can remove it anytime we agree on ever-green browsers.

spacedmonkey reviewed (changes requested):
One small tweak and a question, but otherwise, we are looking good.

merapi committed:
Switch to CommonJS because it’s working, fix JSDoc

ES6 import is failing (repo cause): https://travis-ci.com/github/google/web-stories-wp/jobs/297848579

spacedmonkey commented on assets/src/edit-story/utils/useIntersectionEffect.js:
```diff
@@ -0,0 +1,59 @@
+ * External dependencies
+ */
+import { useEffect } from 'react';
+const IntersectionObserver = require('intersection-observer-polyfill/dist/IntersectionObserver');
```
I don't love this. Did you ever workout why this is happening? Can you explain. If it can't be helped, a comment should be added.

merapi replied:
Yes, it's up to the library author (build config). I hope we will remove all polyfills after we agree on supported browsers.

miina commented on assets/src/edit-story/utils/useIntersectionEffect.js:
```diff
@@ -0,0 +1,59 @@
+ * External dependencies
+ */
+import { useEffect } from 'react';
+const IntersectionObserver = require('intersection-observer-polyfill/dist/IntersectionObserver');
```
Wouldn't `import IntersectionObserver from 'intersection-observer-polyfill';` work?

spacedmonkey replied:
It was changed to fix travis tests it seems - https://github.com/google/web-stories-wp/pull/562/commits/ae012fe31aa7984db6acc3ba812dacf97b78f331

merapi replied:
Nope, it's up to library author. I hope we will remove all polyfills after we agree on supported browsers.

miina replied:
Alright, it seemed to me from [these examples](https://www.npmjs.com/package/intersection-observer-polyfill) that we should be able to use `import`.

merapi replied:
_should_ - we tried that 😉 , just removed the polyfill  🎉

miina commented on assets/src/edit-story/utils/useIntersectionEffect.js:
```diff
@@ -0,0 +1,59 @@
+      const node = ref.current;
+      const usingRoot = 'root' in options;
+      if (usingRoot) {
+        options.root = options.root.current;
```
Any reason for not passing the `options.root.current` directly into the options?

merapi replied:
[here?](https://github.com/google/web-stories-wp/pull/562/files/ae012fe31aa7984db6acc3ba812dacf97b78f331#diff-468f2d4f5a0260eca778fdf38e9a9a3bR238) if yes - I just followed the same convention as in `useResizeEffect`, it takes `ref`, should we only operate on DOM nodes and only take refs if we intend to modify them?

miina replied:
Yes, there. Just that the naming is a bit ambiguous -- `root` seems to be a node by name, however, it's actually passing a ref. I was wondering that maybe you didn't pass the node directly for some specific reason, to skip the `options.root = options.root.current` thing here. There's probably no difference if to pass a ref or a node directly, however, it would be good to match the naming.

merapi replied:
Good call, also this was not quite true
```js
* @param {?Object} options The IntersectionObserver options.
```
Done in https://github.com/google/web-stories-wp/pull/562/commits/715f2a730221c493f47635ac42fc53001fd10b1f

miina commented on assets/src/edit-story/components/library/panes/media/mediaPane.js:
```diff
@@ -182,69 +222,97 @@ function MediaPane(props) {
+              </Column>
+              {hasMore && (
+                <div ref={refContainerFooter}>
+                  {__('Loading', 'web-stories')}
```
Was there any specific UX for this? Probably for a follow-up PR?

spacedmonkey replied:
There is a follow up ticket -  https://github.com/google/web-stories-wp/issues/439

miina commented on assets/src/edit-story/app/media/useMediaReducer/reducer.js:
```diff
@@ -62,27 +76,40 @@ function reducer(state, { type, payload }) {
+      if (mediaType === state.mediaType) return state;
       return {
         ...state,
+        ...resetOnFilterChange,
```
This sounds like a method name, could we rename this to something else? Does this include the default/initial state values basically?

miina replied:
This seems to share quite a few values with `INITIAL_STATE`, perhaps we could use sth like
`FILTER_RESET_STATE` and then the `INITIAL_STATE` could consist of `FILTER_RESET_STATE` + the values unique to that only?

merapi replied:
This should be cleaner: https://github.com/google/web-stories-wp/pull/562/commits/59423a673c935063bf0516dd0ca7cb2b1f2b159d

miina commented on assets/src/edit-story/app/media/mediaProvider.js:
```diff
@@ -31,45 +31,67 @@ import Context from './context';
+    ]
+  );
+
+  const resetAfterUpload = useCallback(() => {
```
Perhaps rename this `resetWithFetch` to make it more general and applicable in the future for other cases, too?

merapi replied:
Done https://github.com/google/web-stories-wp/pull/562/commits/1ffa6d3b720112520e9bdd111c31d9f5907fd3dd

merapi committed:
Remove polyfill 🎉

merapi committed:
Better use of initial state

merapi committed:
Make function name more generic.

miina commented on assets/src/edit-story/app/media/useMediaReducer/reducer.js:
```diff
@@ -53,36 +60,42 @@ function reducer(state, { type, payload }) {
+      if (mediaType === state.mediaType) return state;
       return {
-        ...state,
+        ...INITIAL_STATE,
```
Probably the search term should prevail when switching between media types?

merapi replied:
Correct, in this old UI, it's not so obvious, in Figma it's clear `select` (Filters can be applied on top of a search).
Done in https://github.com/google/web-stories-wp/pull/562/commits/96e2fb4ba24a03d802a4f7a877506afed9083506

merapi committed:
Pass node directly

merapi committed:
Don’t reset search term when switching between types

dvoytenko commented on assets/src/edit-story/components/library/panes/media/mediaPane.js:
```diff
@@ -182,69 +222,97 @@ function MediaPane(props) {
+      setPage({ page: page + 1 });
+    },
+    {
+      root: refContainer.current,
```
This will start as a `null`. Probably not what you want?

merapi replied:
I check it [here](https://github.com/google/web-stories-wp/pull/562/files/96e2fb4ba24a03d802a4f7a877506afed9083506#diff-9e607ff4ff88856913ed42eb4b275859R29) (`node` and `root` if you are using that option)

dvoytenko replied:
That's not enough. The ref starts as null and only set post rendering. I think you should pass `ref` here. Otherwise, by the time you reach [here](https://github.com/google/web-stories-wp/pull/562/files/96e2fb4ba24a03d802a4f7a877506afed9083506#diff-9e607ff4ff88856913ed42eb4b275859R29), the `options.root` is `null`. It's the same reason why you pass the `refContainerFooter` as a ref.

merapi replied:
Right, so I did it in the right way, tested and then broke it with quick "fix"  -> https://github.com/google/web-stories-wp/pull/562#discussion_r392902349

CC @miina so we both learn 😄 
Play around [here](https://codesandbox.io/s/ref-vs-node-tm1t7):
_`refContainer.current` actually works right now, but that's because the component is re-rendered._

dvoytenko replied:
Yeah. That happens a lot, but very very dangerous since the re-rendering is involved. If you want to use the node directly, you will need to include it in the `deps` array. But I think `ref` is a cleaner approach here. Also certainly it messes up a bunch of existing APIs from time to time.

dvoytenko commented on assets/src/edit-story/components/library/panes/media/mediaPane.js:
```diff
@@ -182,69 +222,97 @@ function MediaPane(props) {
+
+  useIntersectionEffect(
+    refContainerFooter,
+    (entry) => {
```
For effect hooks I prefer that the callback and deps go as the last two arguments.

merapi replied:
I was thinking about `options` as a last optional argument, so you can skip them most of the time, but maybe it's a wrong assumption (we are using them here).

dvoytenko replied:
We'll just have to pass `{}` for now. I'm hoping that the React "exhaustive deps" linter will eventually figure out how to do this kind of hooks, and they will really only have two options: first two args are callback and deps, or last two args are. And I think last two args are more likely for them.

dvoytenko commented on assets/src/edit-story/app/api/apiProvider.js:
```diff
@@ -105,10 +105,10 @@ function APIProvider({ children }) {
 
   const getMedia = useCallback(
-    ({ mediaType, searchTerm }) => {
+    ({ mediaType, searchTerm, page }) => {
```
The `page` is such an overloaded term for us. Can we call it a bit more specifically, e.g. `pagingNum` or such?

dvoytenko commented on assets/src/edit-story/app/api/apiProvider.js:
```diff
@@ -118,30 +118,35 @@ function APIProvider({ children }) {
+      return apiFetch({ path: apiPath, parse: false }).then(
+        async (response) => {
+          const json = await response.json();
+          const data = json.map(
```
The resulting `json` is an array? Can we then rename `json` to `jsonArray`?

dvoytenko commented on assets/src/edit-story/app/media/mediaProvider.js:
```diff
@@ -31,45 +31,67 @@ import Context from './context';
     resetFilters,
     setMediaType,
     setSearchTerm,
+    setPage,
```
Ditto: `pagingNum` or such.

dvoytenko commented on assets/src/edit-story/app/media/mediaProvider.js:
```diff
@@ -31,45 +31,67 @@ import Context from './context';
+      fetchMediaStart({ page: p });
+      getMedia({ mediaType, searchTerm, page: p })
+        .then(({ data, headers }) => {
+          const totalPages = parseInt(headers.get('X-WP-TotalPages'));
```
For my info: is this actually WP approach to use response headers for pagination data?

merapi replied:
[Yep...](https://developer.wordpress.org/rest-api/using-the-rest-api/pagination/)

dvoytenko commented on assets/src/edit-story/app/media/useMediaReducer/reducer.js:
```diff
@@ -53,36 +60,43 @@ function reducer(state, { type, payload }) {
       };
     }
 
+    case types.SET_PAGE: {
```
Pls clarify: what does the `SET_PAGE` do comparing to the `FETCH_MEDIA_SUCCESS`? Is this request-vs-response data? That'd seem dangerous to conflate them into a single property, no?

merapi replied:
We now just have `SET_NEXT_PAGE`.

dvoytenko commented on assets/src/edit-story/app/uploader/useUploader.js:
```diff
@@ -72,7 +72,7 @@ function useUploader(refreshLibrary = true) {
     const promise = uploadMedia(file, additionalData);
     if (refreshLibrary) {
-      promise.finally(resetFilters);
+      promise.finally(resetWithFetch);
```
Is there a value in waiting for upload to complete before resetting? Especially if we need to put a placeholder somewhere there?

merapi replied:
I think it will be solved in other PRs (instant preview, local mode).

dvoytenko commented on assets/src/edit-story/components/library/dropzone/index.js:
```diff
@@ -36,8 +36,11 @@ import { ReactComponent as UploadIcon } from './icons/upload.svg';
   min-height: 100%;
+  height: 100%;
+`;
+const OverContent = styled.div`
```
This module has been removed a few days ago.

dvoytenko commented on assets/src/edit-story/components/library/panes/media/mediaPane.js:
```diff
@@ -44,22 +46,26 @@ import {
   grid-gap: 10px;
   grid-template-columns: 1fr 1fr;
+  overflow: auto;
+  padding: 15px 1em 0 1em;
```
Let's use `px` for all paddings. Especially for horizontal padding `em` are a bit confusing.

merapi replied:
I used matching values from the element above. It all will be reworked with a new design waiting in Figma.

dvoytenko commented on assets/src/edit-story/components/library/panes/media/mediaPane.js:
```diff
@@ -182,69 +222,97 @@ function MediaPane(props) {
+  useIntersectionEffect(
+    refContainerFooter,
+    (entry) => {
+      if (!isMediaLoaded || isMediaLoading) return;
```
I think the style requires `{ return }` with new lines and all right now. Though if it doesn't - feel free to leave as is.

dvoytenko commented on assets/src/edit-story/components/library/panes/media/mediaPane.js:
```diff
@@ -182,69 +222,97 @@ function MediaPane(props) {
+    },
+    {
+      root: refContainer.current,
+      rootMargin: '0px 0px 0px 0px',
```
Definitely let's expand right away to an additional viewport height.

dvoytenko commented on assets/src/edit-story/components/library/panes/media/mediaPane.js:
```diff
@@ -182,69 +222,97 @@ function MediaPane(props) {
+              </Column>
+              {hasMore && (
+                <div ref={refContainerFooter}>
+                  {__('Loading', 'web-stories')}
```
Let's add ellipsis here.

dvoytenko commented on assets/src/edit-story/components/library/panes/media/mediaPane.js:
```diff
@@ -182,69 +222,97 @@ function MediaPane(props) {
+                {resources
+                  .filter((_, index) => isEven(index))
+                  .map((resource, i) => (
+                    <MediaElement
```
My first reaction to this is that, most like `loading=lazy` will no longer suffice here. Both because it's not well supported, and also because we'd likely want to add more logic to visible vs not visible state. It's definitely fine to do this over several pull requests, of course. To highlight here: much of the performance tradeoffs will come from the `<video>` where we have more granularity: we can load the video, or just metadata, etc.

And looking at the current `MediaElement`, this would also improve the structure of that component(s). For instance, we can split it into two tiers: `MediaElementWrapper` that takes care of the sizing, focus, hover, etc. And a type-specific `Preview` component that's only rendered when the element is "visible" and can use system resources as it sees fit. This would also make a good place to deal with "poster backfill" and any such issue.

dvoytenko commented on assets/src/edit-story/components/library/panes/media/mediaPane.js:
```diff
@@ -182,69 +222,97 @@ function MediaPane(props) {
+      if (!hasMore) return;
+      if (!entry.isIntersecting) return;
+
+      setPage({ page: page + 1 });
```
Looking more into the use here, I don't see why we need both `page` and `setPage`. I think if we instead introduce `loadNextPage()` method in our context, it'd solve all of these nits: we won't be leaking bulk of the API and the meaning of this will be very clear.

merapi replied:
Agree, but I would call it `setNextPage` because it doesn't load the next page by itself.

merapi committed:
Fix null node

merapi committed:
Apply comments: naming, rootMargin

spacedmonkey reviewed (approved):
This looks good. 👍
